naming-changelog-repartition-topics/kstreams/README.md

```diff
@@ -0,0 +1,93 @@
+Under the hood, these internal topics and state stores have names that, unless provided, are generated by Kafka Streams. Because the generated names include a numeric index based on the position in the topology, changes to the topology can break compatability. This happens because any previously created internal topics and state stores may get orphaned after the topology change.
+
+## Topic and state store generated names
+
+Let's consider a concrete example:
+
+```java
+    KStream<Long, String> inputStream = builder.stream(INPUT_TOPIC, Consumed.with(longSerde, stringSerde))
+            .selectKey((k, v) -> Long.parseLong(v.substring(0, 1)));
+
+    KStream<Long, Long> countStream = inputStream.groupByKey().count().toStream();
+```
+
+Because there is a [`selectKey()`](https://kafka.apache.org/36/javadoc/org/apache/kafka/streams/kstream/KStream.html#selectKey(org.apache.kafka.streams.kstream.KeyValueMapper)) operation that changes the input stream's key, the ensuing 
+[`groupByKey()`](https://kafka.apache.org/36/javadoc/org/apache/kafka/streams/kstream/KStream.html#groupByKey()) operation forces a repartition to an internal topic to make sure the modified keys end up on the correct partition.
+
+Then, the [`count()`](https://kafka.apache.org/36/javadoc/org/apache/kafka/streams/kstream/KGroupedStream.html#count()) aggregation causes Kafka Streams to create a state store and changelog topic (for fault tolerance).
+
+If you were to call [`Topology.describe`](https://kafka.apache.org/36/javadoc/org/apache/kafka/streams/Topology.html#describe()) on this topology, the output would show you the following names for the state store, changelog, and repartition topic:
+
+* state store: `<application.id>-KSTREAM-AGGREGATE-STATE-STORE-0000000002`
+* changelog topic: `<application.id>-KSTREAM-AGGREGATE-STATE-STORE-0000000002-changelog`
+* repartition topic: `<application.id>-KSTREAM-AGGREGATE-STATE-STORE-0000000002-repartition`
+
+The `0000000002` substring in each name is the operation index mentioned above.
+
+Now, let's update the topology by adding a filter:
+
+```java
+    KStream<Long, String> inputStream = builder.stream(INPUT_TOPIC, Consumed.with(longSerde, stringSerde))
+            .selectKey((k, v) -> Long.parseLong(v.substring(0, 1)))
+            .filter((k, v) -> k != 100L);
+
+    KStream<Long, Long> countStream = inputStream.groupByKey().count().toStream();
+```
+
+Since the filter comes before the `count()` aggregation, the number in the state store and changelog / repartition topics will go from `0000000002` to `0000000003`.
+
+The upshot of this is that, when we run the application with the updated topology, the counts will start over from zero rather than start from where they left off.
+
+## Naming state stores and internal topics
+
+How can we reuse state stores and interrnal topics in the face of topology changes? There are a few places in the Kafka Streams DSL where we can control the names of the internal topics and state stores:
+
```

Comment:
Maybe work in the `Named` object in here somewhere that's what you'll use most of the time to name something

Reply:
added a note about this